### PR TITLE
Changed a memory check to catch buffer overflow.

### LINKED ISSUE #909: markdup eats own heap
#### Are you using the latest version of [samtools]
Yes 1.9 release

 #### Please describe your environment.
- OS (run `uname -sr` on Linux/Mac OS or ` wmic os get Caption, Version` on Windows)
- machine architecture (run `uname -m` on Linux/Mac OS or ` wmic os get OSArchitecture` on Windows)
- compiler (run `gcc --version` or `clang --version`)

hgs4c
Linux 3.8.0-44-generic x86_64 Ubuntu
gcc (GCC) 7.2.0
GNU ld (GNU Binutils) 2.24

#### Please specify the steps taken to generate the issue, the command you are running and the relevant output.

```
$ ${samtoolspath}/samtools markdup --reference $ref -O cram -l 151 -T ${base}_align.mkdup -S HG00181_align.sorted.bam ${base}_align.cram
*** glibc detected *** /lustre/scratch115/projects/gdap-wgs/polaris/software/samtools-1.9/samtools: realloc(): invalid next size: 0x000000000199a140 ***
^C
```

ctrl+c applied because of hang. Attempt at valgrind failed because Sanger valgrind is out of date and cannot handle an instruction:

```
==9019== Memcheck, a memory error detector
==9019== Copyright (C) 2002-2011, and GNU GPL'd, by Julian Seward et al.
==9019== Using Valgrind-3.7.0 and LibVEX; rerun with -h for copyright info
==9019== Command: /lustre/scratch115/projects/gdap-wgs/polaris/software/samtools-1.9/samtools markdup --reference /lustre/scratch115/resources/ref/Homo_sapiens/HS38DH/hs38DH.fa -O cram -l 151 -T HG00181_align.mkdup -S HG00181_align.sorted.bam HG00181_align.cram
==9019== 
==9019== Conditional jump or move depends on uninitialised value(s)
==9019==    at 0x50584E0: inflateReset2 (in /lib/x86_64-linux-gnu/libz.so.1.2.3.4)
==9019==    by 0x50585D8: inflateInit2_ (in /lib/x86_64-linux-gnu/libz.so.1.2.3.4)
==9019==    by 0x46FAC1: decompress_peek.constprop.20 (hts.c:144)
==9019==    by 0x471A3D: hts_detect_format (hts.c:209)
==9019==    by 0x477A34: hts_hopen (hts.c:810)
==9019==    by 0x477ECC: hts_open_format (hts.c:426)
==9019==    by 0x455DA5: bam_markdup (bam_markdup.c:981)
==9019==    by 0x64377EC: (below main) (libc-start.c:226)
==9019== 
==9019== Conditional jump or move depends on uninitialised value(s)
==9019==    at 0x50584E0: inflateReset2 (in /lib/x86_64-linux-gnu/libz.so.1.2.3.4)
==9019==    by 0x50585D8: inflateInit2_ (in /lib/x86_64-linux-gnu/libz.so.1.2.3.4)
==9019==    by 0x464141: bgzf_uncompress (bgzf.c:547)
==9019==    by 0x465D07: bgzf_read_block (bgzf.c:572)
==9019==    by 0x46660F: bgzf_read (bgzf.c:987)
==9019==    by 0x47DE98: bam_hdr_read (sam.c:136)
==9019==    by 0x47F190: sam_hdr_read (sam.c:1017)
==9019==    by 0x455FEE: bam_markdup (bam_markdup.c:493)
==9019==    by 0x64377EC: (below main) (libc-start.c:226)
==9019== 
vex amd64->IR: unhandled instruction bytes: 0xF3 0x48 0xF 0xBC 0xC9 0xC1 0xF9 0x3
==9019== valgrind: Unrecognised instruction at address 0x458740.
==9019==    at 0x458740: LZ4_count (lz4.c:319)
==9019==    by 0x45AF45: LZ4_compress_fast_continue (lz4.c:608)
==9019==    by 0x457D92: tmp_file_write_to_file (tmp_file.c:214)
==9019==    by 0x45631F: bam_markdup (bam_markdup.c:787)
==9019==    by 0x64377EC: (below main) (libc-start.c:226)
==9019== Your program just tried to execute an instruction that Valgrind
==9019== did not recognise.  There are two possible reasons for this.
==9019== 1. Your program has a bug and erroneously jumped to a non-code
==9019==    location.  If you are running Memcheck and you just saw a
==9019==    warning about a bad jump, it's probably your program's fault.
==9019== 2. The instruction is legitimate but Valgrind doesn't handle it,
==9019==    i.e. it's Valgrind's fault.  If you think this is the case or
==9019==    you are not sure, please let us know and we'll try to fix it.
==9019== Either way, Valgrind will now raise a SIGILL signal which will
==9019== probably kill your program.
==9019== 
==9019== Process terminating with default action of signal 4 (SIGILL)
==9019==  Illegal opcode at address 0x458740
==9019==    at 0x458740: LZ4_count (lz4.c:319)
==9019==    by 0x45AF45: LZ4_compress_fast_continue (lz4.c:608)
==9019==    by 0x457D92: tmp_file_write_to_file (tmp_file.c:214)
==9019==    by 0x45631F: bam_markdup (bam_markdup.c:787)
==9019==    by 0x64377EC: (below main) (libc-start.c:226)
==9019== 
==9019== HEAP SUMMARY:
==9019==     in use at exit: 10,164,454 bytes in 15,438 blocks
==9019==   total heap usage: 26,057 allocs, 10,619 frees, 3,460,509,326 bytes allocated
==9019== 
==9019== LEAK SUMMARY:
==9019==    definitely lost: 0 bytes in 0 blocks
==9019==    indirectly lost: 0 bytes in 0 blocks
==9019==      possibly lost: 0 bytes in 0 blocks
==9019==    still reachable: 10,164,454 bytes in 15,438 blocks
==9019==         suppressed: 0 bytes in 0 blocks
==9019== Rerun with --leak-check=full to see details of leaked memory
==9019== 
==9019== For counts of detected and suppressed errors, rerun with: -v
==9019== Use --track-origins=yes to see where uninitialised values come from
==9019== ERROR SUMMARY: 23 errors from 2 contexts (suppressed: 2 from 2)
Illegal instruction
```

Test data available on request. (give me a sanger path to copy it to)

### PR DESCRIPTION
**Pull request by whitwham**
Fixes #909.  Additional tidying up of tmp_file based on previous unmerged work.

### CODE REVIEW

File: tmp_file.c [hunk 1]
```diff
@@ -249,20 +252,26 @@ static int tmp_file_write_to_file(tmp_file_t *tmp) {
...
 
             return ret;
         }
     }
 
+
     tmp->ring_index = tmp->ring_buffer + tmp->offset + tmp->input_size;
 
+    if (longest < inbam->l_data) {
+        longest = inbam->l_data;
```
Comment (valeriuo) [inline tmp_file.c]: Is there any use for `longest `?

File: tmp_file.c [hunk 2]
```diff
@@ -249,20 +252,26 @@ static int tmp_file_write_to_file(tmp_file_t *tmp) {
  * Returns 0 on success, a negative number on failure.
```
Comment (valeriuo) [inline tmp_file.c]: Small typo in `Mulitiple `.

File: tmp_file.c (hunk 1 above)
Comment (whitwham) [reply tmp_file.c]: No, that was debugging code.

**Comment by valeriuo** [comment]
Merged in https://github.com/samtools/samtools/commit/f974d2ef38af77c54a8cd37d65f4cdbc7eba7613

**Pull request closed without merging**